Update build and contributing docs (#18223)
umbraco/Umbraco-CMS: .github/BUILD.md

↖️ You can jump to any section by using the "table of contents" button ( ![Table of contents icon](img/tableofcontentsicon.svg) ) above.

## Working with the Umbraco source code

Did you read ["Are you sure"](#are-you-sure)?

[More details about contributing to Umbraco and how to use the GitHub tooling can be found in our guide to contributing.][contribution guidelines]

If you want to run a build without debugging, see [Building from source](#building-from-source) below. This runs the build in the same way it is run on our build servers.

> If you've got this far and are keen to get stuck in helping us fix a big or implement a feature, great! Please read on...

Reviewer: ```suggestion
If you've got this far and are keen to get stuck in helping us fix a bug or implement a feature, great! Please read on...
```

-   [dotnet SDK v9+](https://dotnet.microsoft.com/en-us/download)
-   [Node.js v20+](https://nodejs.org/en/download/)
-   npm v10+ (installed with Node.js)
-   [Git command line](https://git-scm.com/download/)

### Familiarizing yourself with the code

Umbraco is a .NET application using C#. The solution is broken down into multiple projects.  There are several class libraries. The `Umbraco.Web.UI` project is the main project that hosts the back office and login screen. This is the project you will want to run to see your changes.

There are two web projects in the solution with client-side assets based on TypeScript, `Umbraco.Web.UI.Client` and `Umbraco.Web.UI.Login`.

There are a few different ways to work locally when implementing features or fixing issues with the Umbraco CMS. Depending on whether you are working solely on the front-end, solely on the back-end, or somewhere in between, you may find different workflows work best for you.

Here are some suggestions based on how we work on developing Umbraco at HQ.

### First checkout

> When you first clone the source code, rebuild the whole solution via your IDE. You can then start the application via the IDE or the command line and should find everything across front and back-end is built and running.

Reviewer 2: A rebuild is seldomly necessary, seldomly - only when something has gone horribly wrong. Just build.

Reviewer 2: ```suggestion
When you first clone the source code, build the whole solution via your IDE. You can then start the `Umbraco.Web.UI` project via the IDE or the command line and should find everything across front and back-end is built and running.
```

```
cd <solution root>\src\Umbraco.Web.UI
dotnet run --no-build
```

When the page loads in your web browser, you can follow the installer to set up a database for debugging. When complete, you will have an empty Umbraco installation to begin working with. You may also wish to install a [starter kit][https://marketplace.umbraco.com/category/themes-&-starter-kits] to ease your debugging.

### Back-end only changes

If you are working on back-end only features, when switching branches or pulling down the latest from GitHub, you will find the front-end getting rebuilt periodically when you look to build the back-end changes. This can take a while and slow you down. So if for a period of time you don't care about changes in the front-end, you can disable this build step.

Go to `Umbraco.Cms.StaticAssets.csproj` and comment out the following lines of MsBuild by adding a REM statement in front:

```
REM npm ci --no-fund --no-audit --prefer-offline
REM npm run build:for:cms
```

> Just be careful not to include this change in your PR.

Reviewer 2: This is not going to run each time, right? Only when certain folders are missing. I have never commented these out and never had these commands run again.
Encouraging people to update `csproj` will just lead them to commit those changes 😉

Author: Well, in practice I have found it running when I "didn't want it", so I've disabled it myself.

It's a good thing to have for the first experience.  I.e. you just build in your IDE and everything will restore and build.  But I find it triggers if I've:

- Pulled down the latest from GitHub.  Could be there are some FE changes, but I don't really care about them if I've working on a BE only feature.
- Switching branches to review a PR with only BE changes.

And as it can take several minutes, and sometimes fails, this does get in the way of your flow.

Reviewer 2: For sure, that's very unwanted. 

Might be a recent change then as I haven't experienced this. I see why it's there as a convenience but I don't like it 😅 
Anyway, fine to leave this `REM` hint in there then.

Reviewer 2: Just because I was curious: indeed this came in in late November: https://github.com/umbraco/Umbraco-CMS/pull/17527/files#diff-096b40beec2cbe6b2055e325bb6ce85918006bc54fa003508311ffe53bfcfb6aR40-R52

### Front-end only changes

Conversely, if you are working on front-end only, you want to build the back-end once and then run it. Before you do so, update the configuration in `appSettings.json` to add the following under `Umbraco:Cms:Security`:

```
"BackOfficeHost": "http://localhost:5173",
"AuthorizeCallbackPathName": "/oauth_complete",
"AuthorizeCallbackLogoutPathName": "/logout",
"AuthorizeCallbackErrorPathName": "/error"
```

Then run Umbraco from the command line.

```
cd <solution root>\src\Umbraco.Web.UI
dotnet run --no-build
```

In another terminal window, run the following to watch the front-end changes and launch Umbraco using the URL indicated from this task.

```
cd <solution root>\src\Umbraco.Web.UI.Client

> npm run dev:server

Reviewer 2: do an `npm ci` first if it's never had the package restore process run before?

Author: I've added a single reference to when you might have to do this and how you should do it.

```

You'll find as you make changes to the front-end files, the updates will be picked up and your browser refreshed automatically.

> [!NOTE]
> The caching for the back office has been described as 'aggressive' so we often find it's best when making back office changes to [disable caching in the browser (check "Disable cache" on the "Network" tab of developer tools)][disable browser caching] to help you to see the changes you're making.

Whilst most of the backoffice code lives in `Umbraco.Web.UI.Client`, the login screen is in a separate project. If you do any work with that you can build with:

```
cd <solution root>\src\Umbraco.Web.UI.Login

> npm run build

Reviewer 2: do an `npm ci` first if it's never had the package restore process run before?

Author: I've added a single reference to when you might have to do this and how you should do it.

```

### Full-stack changes

If working across both front and back-end, follow both methods and use `dotnet watch`, or re-run `dotnet run` (or `dotnet build` followed by `dotnet run --no-build`) whenever you need to update the back-end code.

Request and response models used by the management APIs are made available client-side as generated code. If you make changes to the management API, you can re-generate the typed client code with:

```
cd <solution root>\src\Umbraco.Web.UI.Client

> npm run generate:server-api-dev

Reviewer 2: do an `npm ci` first if it's never had the package restore process run before?

Author: I've added a single reference to when you might have to do this and how you should do it.

```

> Please also update the `OpenApi.json` file held in the solution by copying and pasting the output from `/umbraco/swagger/management/swagger.json`. Be careful to undo any automated formatting carried out after pasting by the IDE.

Reviewer 2: What happens when I let my IDE autoformat? Can we prevent this?

Author: It doesn't do any harm... it's just that your check-in will show a lot more changes than you've actually made.  It's mostly around collections - the default output from Swagger is `[ ]` but Visual Studio prefers them as `[]`.  So if you let it auto-format you get a lot more lines changed in your PR.

Reviewer 2: Ah okay! I am not too worried about this, putting it in here makes me think something will break. Easy for reviewers to tap the "Hide whitespace" checkbox. 

![image](https://github.com/user-attachments/assets/bf74282f-3d55-4b4f-9f4f-b9c9eb80fea7)

My suggestion would be to leave this hint out. 

```suggestion
Please also update the `OpenApi.json` file held in the solution by copying and pasting the output from `/umbraco/swagger/management/swagger.json`.
```
